### ISSUE TITLE
Default resampling (`0.6mm`) for `sct_deepseg_lesion` QC report distorts lesion softsegs

### ISSUE THREAD

**Issue by valosekj** [spinalcordtoolbox/spinalcordtoolbox#4125]
Newly implemented sagittal lesion QC (PR [#4102](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4102)) works only with binary lesion masks. 

If the lesion mask is non-binary (i.e., soft), the output QC is corrupted:

<img width="1509" alt="image" src="https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/39456460/8a89d065-acdf-4ba7-aa3f-9987030a8e2d">

Example subject: `data.neuro.polymtl.ca/sci-zurich/sub-zh41/ses-01`

Command:

```console
sct_qc -i sub-zh41/ses-01/anat/sub-zh41_ses-01_acq-sag_T2w.nii.gz -s derivatives/labels/sub-zh41/ses-01/anat/sub-zh41_ses-01_acq-sag_T2w_seg-manual.nii.gz -d derivatives/labels/sub-zh41/ses-01/anat/sub-zh41_ses-01_acq-sag_T2w_lesion-manual.nii.gz -p sct_deepseg_lesion -plane sagittal -qc qc
```

I temporarily solved the issue by [lesion mask binarization](https://github.com/ivadomed/model_seg_sci/blob/3bac2896f00bea21fe0dd3d86b2c7d31f029417b/generate_new_lesions/generate_new_lesion.py#L288-L295):

```python
# Binarize new_lesion (sct_qc supports only binary masks)
new_lesion_bin_path = new_lesion_path.replace('.nii.gz', '_bin.nii.gz')
os.system(f'sct_maths -i {new_lesion_path} -bin 0 -o {new_lesion_bin_path}')
# Example: sct_qc -i t2.nii.gz -s t2_seg.nii.gz -d t2_lesion.nii.gz -p sct_deepseg_lesion -plane axial
os.system(f'sct_qc -i {new_target_path} -s {new_sc_path} -d {new_lesion_bin_path} -p sct_deepseg_lesion '
          f'-plane sagittal -qc {args.dir_save.replace("labelsTr", "qc")} -qc-subject {subject_name_out}')
# Remove binarized lesion
os.remove(new_lesion_bin_path)
```

**Comment by joshuacwnewton** [comment]
I can reproduce what you are seeing. Here are some of my initial observations:

- In the original lesion image, I see that the lesion only spans 4 slices. (7-10). I also see that only slice 8 is soft; slices 7, 9, and 10 appear to be binary.
- But, in the QC report, the lesion spans 19 total slices, and all 19 slices appear to be soft.
- If I temporarily turn resampling off, I get an "accurate" QC report (4 slices, only one is soft):

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/65a3daed-6070-4de6-ba47-ed6cdd31baf4)

This appears to be a similar issue as https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3625; In that issue, resampling distorted the segmentation we were trying to display, too.

We temporarily solved that issue by turning resampling off for the centerline QC (https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3626) and so I'm wondering if that would be appropriate here, too, as a short-term solution.

**Comment by jcohenadad** [comment]
The culprit is likely here:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/35b9eb0b5f91bcd7a51dae66d56b0b624c7d78b8/spinalcordtoolbox/reports/slice.py#L300

We should find another way to check if the input image is a segmentation. Or, easier: use linear interpolation for BOTH image and segmentation. It will make the code simpler.

**Comment by jcohenadad** [comment]
Another issue to fix: the number of sagittal slices should remain the same, so it is likely that the 2D 0.6mm resampling was applied to the wrong dimensions. It should be applied to the AP and IS dimensions, when displaying sagittal slices.

**Comment by joshuacwnewton** [comment]
Thank you for the investigation + brainstorming! I'll take a look at this right away. :hearts:

**Comment by joshuacwnewton** [comment]
> The number of sagittal slices should remain the same (as shown [here](https://user-images.githubusercontent.com/16181459/242656927-65a3daed-6070-4de6-ba47-ed6cdd31baf4.png)), so it is likely that the 2D 0.6mm resampling was applied to the wrong dimensions. It should be applied to the AP and IS dimensions, when displaying sagittal slices.

You are absolutely correct. The slicewise resampling code in `Slice` is written in a way that only supports axial slice views, so the resampling gets applied to the wrong axes:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/6db769609baac13eee40eebc818df5eb6a44fc1c/spinalcordtoolbox/reports/slice.py#L289-L291

For a Sagittal view in SAL orientation, it should be `new_size=[p_resample, p_resample, image.dim[6]]` instead.

However, while this does fix the "number of slices" issue, it doesn't fix the opaque background value issue (as expected):

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/f208b52a-3dae-44d0-818d-031f28dd10c6)

**Comment by joshuacwnewton** [comment]
> The culprit is likely here:
>
>  https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/35b9eb0b5f91bcd7a51dae66d56b0b624c7d78b8/spinalcordtoolbox/reports/slice.py#L300-L301

Hm. I'm not sure that binary detection is the issue here. 

Even though it's true that there is 1 slice that is a softseg, in this specific case, 3 of the slices are binary, too. Yet, all 4 slices encounter the opaque background issue, meaning that `is_binary` seems to not be relevant. :thinking: 

----

Digging into things further, I realized that an image is only considered `type_img='seg'` if it is the _last image_ (and 3D):

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/35b9eb0b5f91bcd7a51dae66d56b0b624c7d78b8/spinalcordtoolbox/reports/slice.py#L60-L62

But for this QC, we're providing 2 different segmentations (SC seg for cropping, lesion seg for overlay):

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/35b9eb0b5f91bcd7a51dae66d56b0b624c7d78b8/spinalcordtoolbox/reports/qc.py#L862-L864

Because the SC seg is last, the lesion segmentation is considered type `im`, meaning that:

1. Spline interpolation is applied instead of linear
2. Thresholding doesn't get applied

I'm not sure if one or both of these is the culprit here... 

----

>  Or, easier: use linear interpolation for BOTH image and segmentation. It will make the code simpler.

1. If I try using linear interpolation for `type=im` as suggested (because the lesion seg is considered `type=im`), then the result is definitely better. But, there are still issues:

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/f691390c-bc1f-4995-9b3f-ea4fe162e8e5)

2. If I then try _also_ applying thresholding to the lesion segmentation (by bypassing `type=im`), then I get the following:

![image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/16181459/a0965558-747b-4a04-a0ff-dc1cbe511883)

tl;dr If we choose to resample a softseg, we need to make sure that both linear interpolation _and_ thresholding are applied.

**Comment by jcohenadad** [comment]
Great digging!
